src/open-uri.c

```diff
@@ -282,105 +322,36 @@ handle_open_uri (XdpOpenURI *object,
                                                           arg_parent_window,
                                                           (const char * const *)choices,
                                                           g_variant_builder_end (&opts_builder),
-                                                          &app_chooser_impl_handle,
+                                                          request->id,
                                                           NULL, &error))
     {
       g_dbus_method_invocation_return_gerror (invocation, error);
       return TRUE;
     }
 
-  g_object_set_data_full (G_OBJECT (request), "app-chooser-impl-handle", g_strdup (app_chooser_impl_handle), g_free);
-  register_handle (app_chooser_impl_handle, request);
-
-  g_signal_connect (request, "handle-close", (GCallback)handle_close, request);
-
-  REQUEST_AUTOLOCK (request);
-
-  request_export (request, g_dbus_method_invocation_get_connection (invocation));
-
-  xdp_open_uri_complete_open_uri (object, invocation, request->id);
-  return TRUE;
-}
-
-static void
-update_permissions_store (const char *app_id,const char *content_type, const char *chosen_id)
-{
-  g_autoptr(GError) error = NULL;
-  g_autofree char *latest_chosen_id = NULL;
-  gint latest_chosen_count = 0;
-  g_auto(GStrv) in_permissions = NULL;
-
-  if (get_latest_choice_info (app_id, content_type, &latest_chosen_id, &latest_chosen_count) &&
-      (g_strcmp0 (chosen_id, latest_chosen_id) == 0))
+  impl_request = xdp_impl_request_proxy_new_sync (g_dbus_proxy_get_connection (G_DBUS_PROXY (app_chooser_impl)),
+                                                  G_DBUS_PROXY_FLAGS_NONE,
+                                                  g_dbus_proxy_get_name (G_DBUS_PROXY (app_chooser_impl)),
+                                                  request->id,
+                                                  NULL, &error);
```

Comment:
This GDbusProxy will:
a) Not handle signal emissions on a thread, you need to call something like set_proxy_use_threads (G_DBUS_PROXY (impl_request));
b) Not verify that the signals its handling are *actually* from the backend. In fact, it will even accept such calls from the sandboxed app. So the app could call the portal, then fake a "allowed" backend response.

I think instead of this signal setup we should use regular dbus method calls for the backend. We already have a reference to the requests now, as we're the one passing that in, so we can use that to close the window even though the dialog request has not returned yet.